File: Algorithms/Graph Theory/Snakes and Ladders: The Quickest Way Up.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

class Result
{

    /*
     * Complete the 'quickestWayUp' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. 2D_INTEGER_ARRAY ladders
     *  2. 2D_INTEGER_ARRAY snakes
     */

    public static int quickestWayUp(List<List<int>> ladders, List<List<int>> snakes)
    {
        int[] board = new int[101];
        for (int i = 0; i <= 100; i++)
        {
            board[i] = i;
        }

        foreach (var ladder in ladders)
        {
            int start = ladder[0];
            int end = ladder[1];
            board[start] = end;
        }

        foreach (var snake in snakes)
        {
            int start = snake[0];
            int end = snake[1];
            board[start] = end;
        }

        Queue<(int currentSquare, int moves)> queue = new Queue<(int, int)>();
        bool[] visited = new bool[101];
        queue.Enqueue((1, 0));
        visited[1] = true;

        while (queue.Count > 0)
        {
            var (currentSquare, moves) = queue.Dequeue();

            if (currentSquare == 100)
            {
                return moves;
            }

            for (int roll = 1; roll <= 6; roll++)
            {
                int nextSquare = currentSquare + roll;

                if (nextSquare <= 100 && !visited[nextSquare])
                {
                    visited[nextSquare] = true;
                    queue.Enqueue((board[nextSquare], moves + 1));
                }
            }
        }

        return -1;
    }
}

class Solution
{
    public static void Main(string[] args)
    {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int t = Convert.ToInt32(Console.ReadLine().Trim());

        for (int tItr = 0; tItr < t; tItr++)
        {
            int n = Convert.ToInt32(Console.ReadLine().Trim());

            List<List<int>> ladders = new List<List<int>>();

            for (int i = 0; i < n; i++)
            {
                ladders.Add(Console.ReadLine().TrimEnd().Split(' ').ToList().Select(laddersTemp => Convert.ToInt32(laddersTemp)).ToList());
            }

            int m = Convert.ToInt32(Console.ReadLine().Trim());

            List<List<int>> snakes = new List<List<int>>();

            for (int i = 0; i < m; i++)
            {
                snakes.Add(Console.ReadLine().TrimEnd().Split(' ').ToList().Select(snakesTemp => Convert.ToInt32(snakesTemp)).ToList());
            }

            int result = Result.quickestWayUp(ladders, snakes);

            textWriter.WriteLine(result);
        }

        textWriter.Flush();
        textWriter.Close();
    }
}

File: Algorithms/Graph Theory/Snakes and Ladders: The Quickest Way Up.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::collections::VecDeque;

/*
 * Complete the 'quickestWayUp' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. 2D_INTEGER_ARRAY ladders
 *  2. 2D_INTEGER_ARRAY snakes
 */

fn quickestWayUp(ladders: &[Vec<i32>], snakes: &[Vec<i32>]) -> i32 {
    let mut board: Vec<i32> = (0..101).collect();

    for ladder in ladders {
        let start = ladder[0];
        let end = ladder[1];
        board[start as usize] = end;
    }

    for snake in snakes {
        let start = snake[0];
        let end = snake[1];
        board[start as usize] = end;
    }

    let mut queue: VecDeque<(i32, i32)> = VecDeque::new();
    let mut visited = vec![false; 101];
    queue.push_back((1, 0));
    visited[1] = true;


    while let Some((current_square, moves)) = queue.pop_front() {
        if current_square == 100 {
            return moves;
        }


        for roll in 1..=6 {
            let next_square = current_square + roll;

            if next_square <= 100 && !visited[next_square as usize] {
                visited[next_square as usize] = true;
                queue.push_back((board[next_square as usize], moves + 1));
            }
        }
    }

    -1
}

fn main() {
    let stdin = io::stdin();
    let mut stdin_iterator = stdin.lock().lines();

    let mut fptr = File::create(env::var("OUTPUT_PATH").unwrap()).unwrap();

    let t = stdin_iterator.next().unwrap().unwrap().trim().parse::<i32>().unwrap();

    for _ in 0..t {
        let n = stdin_iterator.next().unwrap().unwrap().trim().parse::<i32>().unwrap();

        let mut ladders: Vec<Vec<i32>> = Vec::with_capacity(n as usize);

        for i in 0..n as usize {
            ladders.push(Vec::with_capacity(2_usize));

            ladders[i] = stdin_iterator.next().unwrap().unwrap()
                .trim_end()
                .split(' ')
                .map(|s| s.to_string().parse::<i32>().unwrap())
                .collect();
        }

        let m = stdin_iterator.next().unwrap().unwrap().trim().parse::<i32>().unwrap();

        let mut snakes: Vec<Vec<i32>> = Vec::with_capacity(m as usize);

        for i in 0..m as usize {
            snakes.push(Vec::with_capacity(2_usize));

            snakes[i] = stdin_iterator.next().unwrap().unwrap()
                .trim_end()
                .split(' ')
                .map(|s| s.to_string().parse::<i32>().unwrap())
                .collect();
        }

        let result = quickestWayUp(&ladders, &snakes);

        writeln!(&mut fptr, "{}", result).ok();
    }
}
